{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1de27ccc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-06-10T10:02:00.032000+01:00",
     "start_time": "2024-06-10T09:01:59.391Z"
    },
    "code_folding": [
     108,
     123,
     159
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "release_transgenic_males"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " \n",
    "function Simulate_timeseries(t,\n",
    "                            starting_population_post_release, \n",
    "                            genotypes_m,     \n",
    "                            genotypes_f,        \n",
    "                            ic_f_vals,\n",
    "                            ic_m_vals,\n",
    "                            fitness_matrix_f,\n",
    "                            fitness_matrix_m,\n",
    "                            params;\n",
    "                            fitness = \"a\"',\n",
    "                            transgenic_genotype = false,\n",
    "                            release_regime = false)\n",
    "    \n",
    "    \n",
    "    \n",
    "    #define storage for n. individuals (f zygote, m zygote, f pupae, m pupae, f adult, m adult)\n",
    "    nF = Array{Float64}(undef,(size(genotypes_f)[1],4,t+1))\n",
    "    nM = Array{Float64}(undef,(size(genotypes_m)[1],4,t+1))\n",
    "\n",
    "    #store starting population from input\n",
    "    nF[:,:,1]=starting_population_post_release[1]\n",
    "    nM[:,:,1]=starting_population_post_release[2]\n",
    "    \n",
    "    #store no. adults in starting pop\n",
    "    genotype_Nf_adult=nF[:,4,1]\n",
    "    genotype_Nm_adult=nM[:,4,1]\n",
    "    #store total number of starting males\n",
    "    starting_Nm = sum(genotype_Nm_adult)\n",
    "    \n",
    "    #For repeat releases, store mask to define which males are released each generation\n",
    "    if release_regime != false\n",
    "        release_mask = [a == transgenic_genotype for a in genotypes_m[:,1]]\n",
    "    end\n",
    "\n",
    "    for gen in 1:t\n",
    "        #For repeat releases, release adult males into population gen t-1 to generate generation t\n",
    "        #And apply the reproductive fitness costs\n",
    "        if (release_regime != false)\n",
    "\n",
    "            genotype_Nm_adult=nM[:,3,gen]\n",
    "            genotype_Nm_adult[release_mask] = genotype_Nm_adult[release_mask] .+ (starting_Nm*release_regime[gen])\n",
    "            nM[:,4,gen] = genotype_Nm_adult\n",
    "        end\n",
    "        \n",
    "        #Production of male and female zygotes\n",
    "        f_gen = params[\"f\"]\n",
    "        n_eggs = genotype_Nf_adult.*f_gen\n",
    "        \n",
    "        n_zygotes_f = Array{Float64}(undef,length(genotypes_f_string))\n",
    "        for o in 1:length(genotypes_f_string)\n",
    "            n_zygotes_f[o] = sum(((ic_f_vals[:,:,o].*n_eggs)').* (genotype_Nm_adult./(sum(genotype_Nm_adult))))\n",
    "        end\n",
    "        n_zygotes_m = Array{Float64}(undef,length(genotypes_m_string))\n",
    "        for o in 1:length(genotypes_m_string)\n",
    "            n_zygotes_m[o] = sum(((ic_m_vals[:,:,o].*n_eggs)').* (genotype_Nm_adult./(sum(genotype_Nm_adult))))\n",
    "        end\n",
    "        nF[:,1,gen+1] = n_zygotes_f\n",
    "        nM[:,1,gen+1] = n_zygotes_m\n",
    "        \n",
    "        #fitess costs are applied to zygotes if applicable\n",
    "        if fitness == \"b\"\n",
    "            n_larvae_F = n_zygotes_f.*fitness_matrix_f[:]\n",
    "            n_larvae_M = n_zygotes_m.*fitness_matrix_m[:]\n",
    "        else\n",
    "            n_larvae_F = n_zygotes_f\n",
    "            n_larvae_M = n_zygotes_m\n",
    "        end\n",
    "\n",
    "        #Apply density dependent costs\n",
    "        n_zygotes = [n_larvae_F,n_larvae_M]\n",
    "        n_pupae = apply_density_dependence(params,n_zygotes)\n",
    "        nF[:,2,gen+1] = n_pupae[1]\n",
    "        nM[:,2,gen+1] = n_pupae[2]\n",
    "\n",
    "        #fitness costs are applied to pupae if applicable\n",
    "        if fitness == \"a\"\n",
    "            n_adults_F = n_pupae[1].*fitness_matrix_f[:]\n",
    "            n_adults_M = n_pupae[2].*fitness_matrix_m[:]\n",
    "        else\n",
    "            n_adults_F = n_pupae[1]\n",
    "            n_adults_M = n_pupae[2]\n",
    "        end\n",
    "        n_adults = [n_adults_F,n_adults_M]\n",
    "        nF[:,3,gen+1] = n_adults[1]\n",
    "        nM[:,3,gen+1] = n_adults[2]\n",
    "\n",
    "        #fitness costs are applied to adults if applicable\n",
    "        n_reproadults_F = n_adults[1]\n",
    "        n_reproadults_M = n_adults[2]\n",
    "        n_reproadults = [n_reproadults_F,n_reproadults_M]\n",
    "        nF[:,4,gen+1] = n_reproadults[1]\n",
    "        nM[:,4,gen+1] = n_reproadults[2]\n",
    "\n",
    "        #Store for no. adults are updated for generation t+1\n",
    "        genotype_Nf_adult= n_reproadults[1]\n",
    "        genotype_Nm_adult= n_reproadults[2]\n",
    "        \n",
    "    end\n",
    "    \n",
    "    output = Dict(\n",
    "        \"simnF\" => nF,\n",
    "        \"simnM\" => nM\n",
    "        )\n",
    "    \n",
    "    return(output)\n",
    "end\n",
    "    \n",
    "function apply_density_dependence(params, \n",
    "                                n_zygotes)\n",
    "    \n",
    "    n_zygotes_sum = sum(sort(n_zygotes[1])) + sum(sort(n_zygotes[2]))\n",
    "    juvenile_survival_prob = params[\"OJ\"]*(params[\"α\"]/(params[\"α\"]+n_zygotes_sum))\n",
    "    n_zygote_dd_f = n_zygotes[1] * juvenile_survival_prob\n",
    "    n_zygote_dd_m = n_zygotes[2] * juvenile_survival_prob\n",
    "\n",
    "    return(n_zygote_dd_f,n_zygote_dd_m)\n",
    "end\n",
    "\"\"\"\n",
    "Initiates the prerelease population of adults. \n",
    "Back-calculates the number of zygotes for a given adult population and parameter values. \n",
    "Returns a Nx8 matrix, where N is the number of genotypes, of numbers of zygote, pupae, adults and repro adults for females and males.\n",
    "\"\"\"\n",
    "function Initiate_prerelease_population_4lifestages(params, genotypes_m, genotypes_f)\n",
    "\n",
    "    \n",
    "    Rm = (params[\"f\"]*params[\"OJ\"])/2\n",
    "    Nf = params[\"α\"]*(Rm-1)/params[\"f\"]\n",
    "\n",
    "    #Simulation has 4 life stages: Zygote, Pupae, Adult, Repro adult\n",
    "    #Define each frequency initially as 0\n",
    "    starting_population_f = Array{Float64}(undef,(length(genotypes_f),4))\n",
    "    starting_population_m = Array{Float64}(undef,(length(genotypes_m),4))\n",
    "    fill!(starting_population_f,0)\n",
    "    fill!(starting_population_m,0)\n",
    "    \n",
    "    #Starting popualtion are all WT\n",
    "    #Assume all pupae, adults and reproadults are fully fertile and there is no mortality between stages\n",
    "    starting_population_f[1,2] = Nf\n",
    "    starting_population_m[1,2] = Nf\n",
    "    starting_population_f[1,3] = Nf\n",
    "    starting_population_m[1,3] = Nf\n",
    "    starting_population_f[1,4] = Nf\n",
    "    starting_population_m[1,4] = Nf\n",
    "    \n",
    "    #Assuming the population is at equilibrium AND the population is pre-release\n",
    "    #No. zygotes\n",
    "    starting_population_f[:,1] =  starting_population_f[:,3] .* params[\"f\"] /2\n",
    "    starting_population_m[:,1] =  starting_population_m[:,3] .* params[\"f\"] /2\n",
    "\n",
    "    return(starting_population_f,starting_population_m)\n",
    "end\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Takes in a Nx8 matrix, where N is the number of genotypes, defining the numbers of genotypes for a population pre-release and \n",
    "returns a Nx8  matrix of numbers of each genotype after release of males. \n",
    "The frequency and genotype of the release males is defined by the input.\n",
    "\"\"\"\n",
    "function release_transgenic_males(starting_population,transgenic_genotype,release_freq,genotypes)\n",
    "    \n",
    "    x=deepcopy(starting_population)\n",
    "    mask = [a == transgenic_genotype for a in genotypes][:]\n",
    "    if sum(mask)>1\n",
    "        print(\"error in assigning trangenic values\")\n",
    "    end\n",
    "    #Here we make the assumption that all initially released individuals are fully fit\n",
    "    x[mask,4] .= sum(starting_population[:,4]).*release_freq\n",
    "    \n",
    "    return(x)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4064b4a2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-06-03T15:50:16.375000+01:00",
     "start_time": "2024-06-03T14:50:16.202Z"
    },
    "code_folding": [
     84,
     109,
     127,
     132
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Simulation_processing_wrapper"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Takes in a simulation output and calculates various summary metrics from the numbers of individuals/gametes. \n",
    "Returns a dictionary containing the same output data with additional entries. \n",
    "\"\"\"\n",
    "function Simulation_processing_wrapper(output,falleles,malleles,genotypes_f,genotypes_m)\n",
    "    alleles_list = [unique([g[i] for g in falleles])[:] for i in 1:length(falleles[1])]\n",
    "    \n",
    "    nF = output[\"simnF\"]\n",
    "    nM = output[\"simnM\"]\n",
    "    t = length(nF[1,1,:])\n",
    "\n",
    "    #Make dataframe header labels\n",
    "    falleles_sym = [join(t) for t in falleles]\n",
    "    malleles_sym = [join(t) for t in malleles]\n",
    "    header_f = Symbol.([join(t) for t in genotypes_f])\n",
    "    header_m = Symbol.([join(t) for t in genotypes_m])\n",
    "\n",
    "    #Calculate relative no. females\n",
    "    pupae = [sum(nF[:,2,i]) for i in 1:t]\n",
    "    adult = [sum(nF[:,3,i]) for i in 1:t]\n",
    "    reproadult = [sum(nF[:,4,i]) for i in 1:t]\n",
    "    total_Nf = DataFrame(hcat(pupae,adult,reproadult),[\"pupae\",\"adult\",\"reproadult\"])\n",
    "    relpupae = pupae./pupae[1]\n",
    "    reladult = adult./adult[1]\n",
    "    relreproadult = reproadult./reproadult[1]\n",
    "    relative_Nf = DataFrame(hcat(relpupae,reladult,relreproadult),[\"pupae\",\"adult\",\"reproadult\"])\n",
    "\n",
    "    #Calculate allele frequencies for zygotes\n",
    "    genotypenumber_zygote_female = DataFrame(Transpose(nF[:,1,:]), header_f)\n",
    "    genotypenumber_zygote_male = DataFrame(Transpose(nM[:,1,:]), header_m)\n",
    "    genotype_freqzf_store = freq_from_no_df(genotypenumber_zygote_female)\n",
    "    genotype_freqzm_store = freq_from_no_df(genotypenumber_zygote_male)\n",
    "    genotype_freqz_store = (genotype_freqzf_store.+genotype_freqzm_store)./2;\n",
    "    allele_freqzf = Find_allele_frequencies_from_genotypes_df(genotypes_f,genotype_freqzf_store,alleles_list);\n",
    "    allele_freqzm = Find_allele_frequencies_from_genotypes_df(genotypes_m,genotype_freqzm_store,alleles_list);\n",
    "    allele_freqz = (allele_freqzf .+ allele_freqzm) ./2\n",
    "\n",
    "    #Calculate allele frequencies for adult\n",
    "    genotypenumber_adult_female = DataFrame(Transpose(nF[:,3,:]), header_f)\n",
    "    genotypenumber_adult_male = DataFrame(Transpose(nM[:,3,:]), header_m)\n",
    "    genotype_freqaf_store = freq_from_no_df(genotypenumber_adult_female)\n",
    "    genotype_freqam_store = freq_from_no_df(genotypenumber_adult_male)\n",
    "    genotype_freqa_store = (genotype_freqaf_store.+genotype_freqam_store)./2;\n",
    "    allele_freqaf = Find_allele_frequencies_from_genotypes_df(genotypes_f,genotype_freqaf_store,alleles_list);\n",
    "    allele_freqam = Find_allele_frequencies_from_genotypes_df(genotypes_m,genotype_freqam_store,alleles_list);\n",
    "    allele_freqa = (allele_freqaf .+ allele_freqam) ./2\n",
    "    \n",
    "    #Calculate allele frequencies for zygotes\n",
    "    genotypenumber_reproadult_female = DataFrame(Transpose(nF[:,4,:]), header_f)\n",
    "    genotypenumber_reproadult_male = DataFrame(Transpose(nM[:,4,:]), header_m)\n",
    "    genotype_freqrf_store = freq_from_no_df(genotypenumber_reproadult_female)\n",
    "    genotype_freqrm_store = freq_from_no_df(genotypenumber_reproadult_male)\n",
    "    genotype_freqr_store = (genotype_freqrf_store.+genotype_freqrm_store)./2;\n",
    "    allele_freqrf = Find_allele_frequencies_from_genotypes_df(genotypes_f,genotype_freqrf_store,alleles_list);\n",
    "    allele_freqrm = Find_allele_frequencies_from_genotypes_df(genotypes_m,genotype_freqrm_store,alleles_list);\n",
    "    allele_freqr = (allele_freqrf .+ allele_freqrm) ./2\n",
    "\n",
    "    #save in a dictionary\n",
    "    data_dict = Dict(\n",
    "\n",
    "        \"simnF\" => nF,\n",
    "        \"simnM\" => nM,\n",
    "        \"number_females\" => total_Nf,\n",
    "        \"relative_number_females\" => relative_Nf,\n",
    "        \"genotypenumber_zygote_female\" => genotypenumber_zygote_female,\n",
    "        \"genotypenumber_zygote_male\" => genotypenumber_zygote_male,\n",
    "        \"genotypefreq_zygote_female\" => genotype_freqzf_store,\n",
    "        \"genotypefreq_zygote_male\" => genotype_freqzm_store,\n",
    "        \"genotypefreq_zygote\" => genotype_freqz_store,\n",
    "        \"allelefreq_zygote\" => allele_freqz,\n",
    "        \"allelefreq_zygote_female\" => allele_freqzf,\n",
    "        \"allelefreq_zygote_male\" => allele_freqzm,\n",
    "        \n",
    "        \"genotypenumber_adult_female\" => genotypenumber_adult_female,\n",
    "        \"genotypenumber_adult_male\" => genotypenumber_adult_male,\n",
    "        \"genotypefreq_adult_female\" => genotype_freqaf_store,\n",
    "        \"genotypefreq_adult_male\" => genotype_freqam_store,\n",
    "        \"genotypefreq_adult\" => genotype_freqa_store,\n",
    "        \"allelefreq_adult\" => allele_freqa,\n",
    "        \"allelefreq_adult_female\" => allele_freqaf,\n",
    "        \"allelefreq_adult_male\" => allele_freqam,\n",
    "        \n",
    "        \"genotypenumber_reproadult_female\" => genotypenumber_reproadult_female,\n",
    "        \"genotypenumber_reproadult_male\" => genotypenumber_reproadult_male,\n",
    "        \"genotypefreq_reproadult_female\" => genotype_freqrf_store,\n",
    "        \"genotypefreq_reproadult_male\" => genotype_freqrm_store,\n",
    "        \"genotypefreq_reproadult\" => genotype_freqr_store,\n",
    "        \"allelefreq_reproadult\" => allele_freqr,\n",
    "        \"allelefreq_reproadult_female\" => allele_freqrf,\n",
    "        \"allelefreq_reproadult_male\" => allele_freqrm,\n",
    "    );\n",
    "    return(data_dict)\n",
    "end\n",
    "\n",
    "\n",
    "\"\"\"Calculate the row-wise frequency\"\"\"\n",
    "function freq_from_no_df(genotype_N_store)\n",
    "    genotype_N_store./sum(sort(Matrix(genotype_N_store),dims=2),dims=2)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the allele frequency.\n",
    "\"\"\"\n",
    "function Find_allele_frequencies_from_genotypes(genotypes,freq_genotype,alleles_list)\n",
    "    \"\"\"Extract the allele frequencies from the genotype frequencies\"\"\"\n",
    "    allele_freqs = Vector{Float64}()\n",
    "    for allele in vcat(alleles_list...)\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if (allele in genotypes[i][:,1]) & (allele in genotypes[i][:,2])\n",
    "                allele_freq = allele_freq+freq_genotype[i]\n",
    "            elseif (allele in genotypes[i][:,1]) | (allele in genotypes[i][:,2])\n",
    "                allele_freq = allele_freq+(freq_genotype[i]/2)\n",
    "            end\n",
    "        end\n",
    "       push!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "function Find_allele_frequencies_from_genotypes_df(genotypes,genotype_freq_store,alleles_list)\n",
    "    header = Symbol.(vcat(alleles_list...))\n",
    "    Make_dataframe([Find_allele_frequencies_from_genotypes(genotypes,row,alleles_list) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Takes in the genotype frequencies of a population at a single time point and \n",
    "calculated the frequency of each haploid genotype (chromosome).\n",
    "\"\"\"\n",
    "function Find_chromosome_frequencies_from_genotypes(freq_genotype,genotypes,alleles)\n",
    "    \"\"\"Extract the allele frequencies from the genotype frequencies\"\"\"\n",
    "    allele_freqs = []\n",
    "    for allele in alleles\n",
    "        allele_freq = 0\n",
    "        for i in 1:length(genotypes)\n",
    "            if ((allele == genotypes[i][:,1]) | (allele == genotypes[i][:,2]))\n",
    "                if genotypes[i][:,1] == genotypes[i][:,2]\n",
    "                    allele_freq = allele_freq+(freq_genotype[i])\n",
    "                else\n",
    "                    allele_freq = allele_freq+(freq_genotype[i]/ 2)\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "       append!(allele_freqs,allele_freq)\n",
    "    end\n",
    "    return(allele_freqs)\n",
    "end\n",
    "function Find_chromosome_frequencies_from_genotypes_df(genotype_freq_store,genotypes,alleles)\n",
    "    header = Symbol.(alleles)\n",
    "    Make_dataframe([Find_chromosome_frequencies_from_genotypes(row,genotypes,alleles) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "function Find_correlation(genotypes,alleles,genotype_freq,alleles_list,A,B)\n",
    "    AB_alleles = vcat(alleles_list...)\n",
    "    \n",
    "    chrom_freq = Find_chromosome_frequencies_from_genotypes(genotype_freq,genotypes,alleles)\n",
    "    allele_freq = Find_allele_frequencies_from_genotypes(genotypes,genotype_freq,alleles_list)\n",
    "\n",
    "    ab_freq = sum(chrom_freq[[A in x for x in alleles] .& [B in x for x in alleles]])\n",
    "    a_freq = allele_freq[findall(x->x==A,AB_alleles)[1]]\n",
    "    b_freq = allele_freq[findall(x->x==B,AB_alleles)[1]]\n",
    "\n",
    "    if a_freq > 1.0\n",
    "        a_freq = round(a_freq,digits=15)\n",
    "    end\n",
    "    if b_freq > 1.0\n",
    "        b_freq = round(b_freq,digits=15)\n",
    "    end\n",
    "\n",
    "    if (float(b_freq) > 0.0)\n",
    "        cor = (ab_freq - (a_freq*b_freq))/sqrt(float(a_freq*(1-a_freq)*b_freq*(1-b_freq)))\n",
    "    else\n",
    "        cor = 0.0\n",
    "    end\n",
    "    return(cor)\n",
    "end\n",
    "function Find_correlation_df(genotypes,alleles,genotype_freq_store,alleles_list,A,B)\n",
    "    header = [\"Correlation\"]\n",
    "    Make_dataframe([Find_correlation(genotypes,alleles,row,alleles_list,A,B) for row = eachrow(genotype_freq_store)],header)\n",
    "end\n",
    "\n",
    "\n",
    "\"\"\"Processes a matrix into a dataframe\"\"\"\n",
    "function Make_dataframe(store, header)\n",
    "    df = DataFrame(Transpose(hcat(store...)),header)\n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8865baa8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-06-14T11:26:58.713000+01:00",
     "start_time": "2024-06-14T10:26:58.081Z"
    },
    "code_folding": [
     0,
     99
    ]
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "find_release_rate_required_nopm (generic function with 1 method)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "function find_release_rate_required(n_releases,\n",
    "                                    release_freq,\n",
    "                                    threshold,\n",
    "                                    pdict,\n",
    "                                    inputs;\n",
    "                                    release_genotype=[\"A1\" \"A2\"],\n",
    "                                    param_mod=false,\n",
    "                                    param_label=false,\n",
    "                                    t=100,\n",
    "                                    increment = 10,\n",
    "                                    Rm=6,\n",
    "                                    accuracy=0.001,\n",
    "                                    fitness = \"a\")\n",
    "\n",
    "    checker = false\n",
    "    lim = 1000\n",
    "    \n",
    "    genotypes_f,\n",
    "    genotypes_m,\n",
    "    maternal_gt,\n",
    "    paternal_gt,\n",
    "    fitness_f,\n",
    "    fitness_m = inputs\n",
    "    \n",
    "    opt_subs_pdict = deepcopy(pdict)\n",
    "    \n",
    "    #mod params\n",
    "    if param_mod != false\n",
    "        for param in 1:length(param_mod)\n",
    "            opt_subs_pdict[param_label[param]]=param_mod[param]\n",
    "        end\n",
    "    end\n",
    "    opt_subs_list = [(param,opt_subs_pdict[param]) for param in keys(pdict)];\n",
    "    \n",
    "    #Fitness\n",
    "    fitness_f_vals = Array{Float64}(fitness_f.subs(opt_subs_list))\n",
    "    fitness_m_vals = Array{Float64}(fitness_m.subs(opt_subs_list))\n",
    "    \n",
    "    maternal_gt_vals = Array{Float64}(maternal_gt.subs(opt_subs_list))\n",
    "    paternal_gt_vals = Array{Float64}(paternal_gt.subs(opt_subs_list))\n",
    "\n",
    "    ic_f_vals,ic_m_vals  = make_ics(maternal_gt_vals,paternal_gt_vals);\n",
    "\n",
    "    \n",
    "    #Dynamics params\n",
    "    params = Dict()\n",
    "    Nf_eq = 1000.0\n",
    "    params[\"OJ\"] = 1/10\n",
    "    params[\"f\"] = (2*Rm)/params[\"OJ\"]\n",
    "    params[\"α\"] = params[\"f\"]*Nf_eq/(Rm-1)\n",
    "    #params[\"fm\"] = 0\n",
    "    \n",
    "    while checker == false\n",
    "        \n",
    "        release_regime = repeat([0.0],t)\n",
    "        for i in 1:n_releases\n",
    "            release_regime[i] = release_freq\n",
    "        end\n",
    "\n",
    "        #Set up release\n",
    "        starting_population_Nf,starting_population_Nm = Initiate_prerelease_population_4lifestages(params, genotypes_m,genotypes_f)\n",
    "        if release_regime == false\n",
    "            starting_population_post_release_m = release_transgenic_males(starting_population_Nm,release_genotype,release_freq,genotypes_m)\n",
    "        else\n",
    "            starting_population_post_release_m = starting_population_Nm\n",
    "        end\n",
    "        starting_population_post_release = [starting_population_Nf,starting_population_post_release_m]\n",
    "\n",
    "        output = Simulate_timeseries(t,             #number of generations\n",
    "                            starting_population_post_release, #[1] = female, [2] = male\n",
    "                            genotypes_m,     #list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "                            genotypes_f,     #list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]) \n",
    "                            ic_f_vals,\n",
    "                            ic_m_vals,\n",
    "                            fitness_f_vals,\n",
    "                            fitness_m_vals,\n",
    "                            params,\n",
    "                            release_regime = release_regime,\n",
    "                            transgenic_genotype = release_genotype,\n",
    "                            fitness = fitness)\n",
    "        \n",
    "        output = Simulation_processing_wrapper(output,falleles,malleles,genotypes_f,genotypes_m);\n",
    "        if ((minimum(output[\"relative_number_females\"].adult) > threshold) & (release_freq<lim))\n",
    "            release_freq = release_freq+increment\n",
    "        elseif release_freq>=lim\n",
    "            checker = true\n",
    "            return(release_freq)\n",
    "        else\n",
    "            if increment != accuracy\n",
    "                release_freq = release_freq-increment\n",
    "                increment = increment/10\n",
    "            else\n",
    "                checker = true\n",
    "                return(release_freq)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end\n",
    "\n",
    "function find_release_rate_required_nopm(n_releases,\n",
    "                                    release_freq,\n",
    "                                    threshold,\n",
    "                                    pdict,\n",
    "                                    input_vals;\n",
    "                                    release_genotype=[\"A1\" \"A2\"],\n",
    "                                    t=100,\n",
    "                                    increment = 10,\n",
    "                                    Rm=6,\n",
    "                                    accuracy=0.001,\n",
    "                                    fitness = \"a\")\n",
    "\n",
    "    checker = false\n",
    "    lim = 1000\n",
    "    \n",
    "    genotypes_f,\n",
    "    genotypes_m,\n",
    "    ic_f_vals,\n",
    "    ic_m_vals,\n",
    "    fitness_f_vals,\n",
    "    fitness_m_vals = input_vals\n",
    "    \n",
    "    #Dynamics params\n",
    "    params = Dict()\n",
    "    Nf_eq = 1000.0\n",
    "    params[\"OJ\"] = 1/10\n",
    "    params[\"f\"] = (2*Rm)/params[\"OJ\"]\n",
    "    params[\"α\"] = params[\"f\"]*Nf_eq/(Rm-1)\n",
    "    #params[\"fm\"] = 0\n",
    "    \n",
    "    while checker == false\n",
    "        \n",
    "        release_regime = repeat([0.0],t)\n",
    "        for i in 1:n_releases\n",
    "            release_regime[i] = release_freq\n",
    "        end\n",
    "\n",
    "        #Set up release\n",
    "        starting_population_Nf,starting_population_Nm = Initiate_prerelease_population_4lifestages(params, genotypes_m,genotypes_f)\n",
    "        if release_regime == false\n",
    "            starting_population_post_release_m = release_transgenic_males(starting_population_Nm,release_genotype,release_freq,genotypes_m)\n",
    "        else\n",
    "            starting_population_post_release_m = starting_population_Nm\n",
    "        end\n",
    "        starting_population_post_release = [starting_population_Nf,starting_population_post_release_m]\n",
    "\n",
    "        output = Simulate_timeseries(t,             #number of generations\n",
    "                            starting_population_post_release, #[1] = female, [2] = male\n",
    "                            genotypes_m,     #list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]\n",
    "                            genotypes_f,     #list of genotypes, each entry = [\"A1\" \"A2\"; \"B1\" \"B2\"]) \n",
    "                            ic_f_vals,\n",
    "                            ic_m_vals,\n",
    "                            fitness_f_vals,\n",
    "                            fitness_m_vals,\n",
    "                            params,\n",
    "                            release_regime = release_regime,\n",
    "                            transgenic_genotype = release_genotype,\n",
    "                            fitness = fitness)\n",
    "        \n",
    "        output = Simulation_processing_wrapper(output,falleles,malleles,genotypes_f,genotypes_m);\n",
    "        if ((minimum(output[\"relative_number_females\"].adult) > threshold) & (release_freq<lim))\n",
    "            release_freq = release_freq+increment\n",
    "        elseif release_freq>=lim\n",
    "            checker = true\n",
    "            return(release_freq)\n",
    "        else\n",
    "            if increment != accuracy\n",
    "                release_freq = release_freq-increment\n",
    "                increment = increment/10\n",
    "            else\n",
    "                checker = true\n",
    "                return(release_freq)\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "733ccde4",
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "function Simulate_single_release_example(opt_subs_list,inputs,t,release_freq,release_genotype; release_regime = false,Rm = 6,fitness = \"a\")\n",
    "\n",
    "    genotypes_f,\n",
    "    genotypes_m,\n",
    "    maternal_gt,\n",
    "    paternal_gt,\n",
    "    fitness_f,\n",
    "    fitness_m = inputs\n",
    "    \n",
    "    #Fitness\n",
    "    fitness_f_vals = Array{Float64}(fitness_f.subs(opt_subs_list))\n",
    "    fitness_m_vals = Array{Float64}(fitness_m.subs(opt_subs_list))\n",
    "    \n",
    "    maternal_gt_vals = Array{Float64}(maternal_gt.subs(opt_subs_list))\n",
    "    paternal_gt_vals = Array{Float64}(paternal_gt.subs(opt_subs_list))\n",
    "\n",
    "    ic_f_vals,ic_m_vals  = make_ics(maternal_gt_vals,paternal_gt_vals);\n",
    "    \n",
    "    # Define pop dynamics params\n",
    "    params = Dict()\n",
    "    Nf_eq = 1000.0\n",
    "    params[\"OJ\"] = 1/10\n",
    "    params[\"f\"] = (2*Rm)/params[\"OJ\"]\n",
    "    params[\"α\"] = params[\"f\"]*Nf_eq/(Rm-1)\n",
    "    params[\"fm\"] = 0\n",
    "\n",
    "    #Set up release\n",
    "    starting_population_Nf,starting_population_Nm = Initiate_prerelease_population_4lifestages(params, genotypes_m,genotypes_f)\n",
    "    if release_regime == false\n",
    "        starting_population_post_release_m = release_transgenic_males(starting_population_Nm,release_genotype,release_freq,genotypes)\n",
    "    else\n",
    "        starting_population_post_release_m = starting_population_Nm\n",
    "    end\n",
    "    starting_population_post_release = [starting_population_Nf,starting_population_post_release_m]\n",
    "\n",
    "    output = Simulate_timeseries(t, \n",
    "                    starting_population_post_release, #[1] = female, [2] = male\n",
    "                    genotypes_m,\n",
    "                    genotypes_f,\n",
    "                    ic_f_vals,\n",
    "                    ic_m_vals,\n",
    "                    fitness_f_vals,\n",
    "                    fitness_m_vals,\n",
    "                    params,\n",
    "                    release_regime = release_regime,\n",
    "                    transgenic_genotype = release_genotype,\n",
    "                    fitness = fitness)\n",
    "    \n",
    "    out = Simulation_processing_wrapper(output,falleles,malleles,genotypes_f,genotypes_m);\n",
    "    return(out)\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df8278de",
   "metadata": {
    "code_folding": [
     3,
     17
    ]
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Sorts the allele string so order is in numerical order\n",
    "\"\"\"\n",
    "function sort_allele(unsorted_allele)\n",
    "    n1 = parse(Int64,unsorted_allele[1][2:end])\n",
    "    n2 = parse(Int64,unsorted_allele[2][2:end])\n",
    "    if n1 > n2\n",
    "        sorted_allele = [unsorted_allele[2,:]; unsorted_allele[1,:]]\n",
    "    else\n",
    "        sorted_allele = [unsorted_allele[1,:]; unsorted_allele[2,:]]\n",
    "    end\n",
    "    return(sorted_allele)\n",
    "end\n",
    "\n",
    "\"\"\"\n",
    "Sorts the genotype string so order is in numerical order and A > B > C etc...\n",
    "\"\"\"\n",
    "function sort_genotype(unsorted_genotype; n=1)\n",
    "    \n",
    "    function assign(sorted_allele,unsorted_allele,unsorted_genotype)\n",
    "        sorted_genotype = Array{String}(undef,(length(loci),2))\n",
    "        if sorted_allele == unsorted_allele\n",
    "            sorted_genotype = unsorted_genotype\n",
    "        else\n",
    "            sorted_genotype[:,1] = unsorted_genotype[:,2]\n",
    "            sorted_genotype[:,2] = unsorted_genotype[:,1]\n",
    "        end\n",
    "        return(sorted_genotype)\n",
    "    end\n",
    "    \n",
    "    if unsorted_genotype[n,1] != unsorted_genotype[n,2]\n",
    "        sorted_genotype = assign(sort_allele(unsorted_genotype[n,:]),unsorted_genotype[n,:],unsorted_genotype)\n",
    "    else\n",
    "        if n < length(loci)\n",
    "            sorted_genotype = sort_genotype(unsorted_genotype,n=n+1)\n",
    "        else\n",
    "            sorted_genotype = unsorted_genotype\n",
    "        end\n",
    "    end\n",
    "    return(sorted_genotype)\n",
    "end\n",
    "\n",
    "\n",
    "function make_ics(allele_matrix_f,m_allele_matrix)\n",
    "\n",
    "    n_alleles = Integer(length(m_allele_matrix[1,:])/2)\n",
    "    m_allele_matrix_m = m_allele_matrix[:,1:n_alleles]\n",
    "    m_allele_matrix_f = m_allele_matrix[:,n_alleles+1:end]\n",
    "    \n",
    "    #production of offspring\n",
    "    ih_f = Array{Float64}(undef,length(genotypes_f),length(genotypes_m),length(genotypes_f))\n",
    "    ih_m = Array{Float64}(undef,length(genotypes_f),length(genotypes_m),length(genotypes_m))\n",
    "    fill!(ih_f, 0.0)\n",
    "    fill!(ih_m, 0.0);\n",
    "\n",
    "    for f in 1:length(genotypes_f) #mothers\n",
    "        for m in 1:length(genotypes_m) #fathers\n",
    "            for e in 1:length(falleles) #autosomal allele from egg\n",
    "                for s in 1:length(falleles) #autosomal allele from egg\n",
    "                    #X\n",
    "                    e_allele = falleles[e]\n",
    "                    s_allele = falleles[s]\n",
    "                    offspring_genotype = sort_genotype(hcat(e_allele,s_allele))\n",
    "                    genotypes_f\n",
    "                    o = findall(x->x==offspring_genotype,genotypes_f)[1]\n",
    "                    ih_f[f,m,o] = ih_f[f,m,o] + (allele_matrix_f[f,e] * m_allele_matrix_f[m,s])\n",
    "                end\n",
    "\n",
    "                for s in 1:length(malleles) #autosomal allele from egg\n",
    "                    #Y\n",
    "                    e_allele = falleles[e]\n",
    "                    s_allele = malleles[s]\n",
    "                    offspring_genotype = sort_genotype(hcat(e_allele,s_allele))\n",
    "                    o = findall(x->x==offspring_genotype,genotypes_m)[1]\n",
    "                    ih_m[f,m,o] = ih_m[f,m,o] + (allele_matrix_f[f,e] * m_allele_matrix_m[m,s])\n",
    "                end\n",
    "            end\n",
    "        end\n",
    "    end\n",
    "\n",
    "    return(ih_f,ih_m)\n",
    "end\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Julia 1.9.3",
   "language": "julia",
   "name": "julia-1.9"
  },
  "language_info": {
   "file_extension": ".jl",
   "mimetype": "application/julia",
   "name": "julia",
   "version": "1.9.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
